probe-rs/src/debug/registers.rs

```diff
@@ -1,162 +1,243 @@
-    /// [DWARF](https://dwarfstd.org)
-    pub fn set_frame_pointer(&mut self, value: Option<u64>) {
-        let register_address = self.register_description.frame_pointer().id.0 as u32;
-
-        if let Some(value) = value {
-            self.values.insert(register_address, value);
-        } else {
-            self.values.remove(&register_address);
-        }
+    pub fn get_frame_pointer(&self) -> Option<&DebugRegister> {
+        self.0.iter().find(|debug_register| {
+            debug_register.id == debug_register.register_file.frame_pointer().id
+        })
     }
 
     /// Get the program counter.
-    pub fn get_program_counter(&self) -> Option<u64> {
-        let reg_num = self.register_description.program_counter().id.0 as u32;
-
-        self.values.get(&reg_num).copied()
-    }
-
-    /// Set the program counter.
-    pub fn set_program_counter(&mut self, value: Option<u64>) {
-        let register_address = self.register_description.program_counter().id.0 as u32;
-
-        if let Some(value) = value {
-            self.values.insert(register_address, value);
-        } else {
-            self.values.remove(&register_address);
-        }
+    pub fn get_program_counter(&self) -> Option<&DebugRegister> {
+        self.0.iter().find(|debug_register| {
+            debug_register.id == debug_register.register_file.program_counter().id
+        })
     }
 
     /// Get the stack pointer.
-    pub fn get_stack_pointer(&self) -> Option<u64> {
-        let reg_num = self.register_description.stack_pointer().id.0 as u32;
-
-        self.values.get(&reg_num).copied()
-    }
-
-    /// Set the stack pointer.
-    pub fn set_stack_pointer(&mut self, value: Option<u64>) {
-        let register_address = self.register_description.stack_pointer().id.0 as u32;
-
-        if let Some(value) = value {
-            self.values.insert(register_address, value);
-        } else {
-            self.values.remove(&register_address);
-        }
+    pub fn get_stack_pointer(&self) -> Option<&DebugRegister> {
+        self.0.iter().find(|debug_register| {
+            debug_register.id == debug_register.register_file.stack_pointer().id
+        })
     }
 
     /// Get the return address.
-    pub fn get_return_address(&self) -> Option<u64> {
-        let reg_num = self.register_description.return_address().id.0 as u32;
-
-        self.values.get(&reg_num).copied()
-    }
-
-    /// Set the return address.
-    pub fn set_return_address(&mut self, value: Option<u64>) {
-        let register_address = self.register_description.return_address().id.0 as u32;
-
-        if let Some(value) = value {
-            self.values.insert(register_address, value);
-        } else {
-            self.values.remove(&register_address);
-        }
+    pub fn get_return_address(&self) -> Option<&DebugRegister> {
+        self.0.iter().find(|debug_register| {
```

Comment:
These patterns are probably fine given the suspected access patterns, but if they get called a lot it might be worth considering a HashMap lookup by ID to avoid searching the entire set every time.

Reply:
It was a trade-off between that, and having to create a way to remember the sort order of the registers (which is lost when in a HashMap).

Reply:
Makes sense.  I'm not sure this is a problem and it can always be reworked later without issue as it's behind an API.